### PR TITLE
Oy2-9910 Add PCR-AD Measure

### PR DESCRIPTION
**Pull request by jphechter**
## Purpose

Build and implement the PCR-AD measure.

#### Linked Issues to Close

[Oy2-9910](https://qmacbis.atlassian.net/browse/OY2-9910?atlOrigin=eyJpIjoiMzk0ZTVjNzQxMDVjNGVjNGE5NjMxNzRhODkxMTc4YTAiLCJwIjoiaiJ9)

## Approach

This measure captures NDRs in a completely unique pattern requiring that we build an alternative to the `QMR.Rate` component. This component is referred to as MultiRate in the code (but the name is completely flexible) as it uses different combinations of field values to compute its rates.

The data captured in this component follows the same rules as data captured in the other sections (non-zero denominators, numerator cannot be greater than denominator, etc), and therefore must have unique validations due to the unique data structure.

Things added in addition to new measure files:
- MultiRate component file
- MultiRate tests
- PCR-AD specific validations that check the data structure of the MultiRate component

## Assorted Notes/Considerations

This component has been specifically built for PCR-AD, but it is still data-driven and could be used in another measure with some minimal work. Things like validation functions could be moved into a more globally accessible location.

#### Pull Request Creator Checklist

- [ ] This PR has an associated issue or issues.
- [ ] The associated issue(s) are linked above.
- [ ] This PR meets all acceptance criteria for those issues.
- [ ] This PR and linked issue(s) are adequately documented
- [ ] This PR and linked issues(s) are a complete description of the changeset; an individual or team should be able to understand the issue(s) and changes by reading through this PR and it's links, with no further interaction.
- [ ] Someone has been assigned this PR.
- [ ] At least one person has been marked as reviewer on this PR.

#### Pull Request Reviewer/Assignee Checklist

- [ ] This PR has an associated issue or issues.
- [ ] The associated issue(s) are linked above.
- [ ] This PR meets all acceptance criteria for those issues.
- [ ] This PR and linked issue(s) are adequately documented
- [ ] This PR and linked issues(s) are a complete description of the changeset; an individual or team should be able to understand the issue(s) and changes by reading through this PR and it's links, with no further interaction.

### CODE REVIEW

File: services/ui-src/src/utils/form.ts [hunk 1]
```diff
@@ -6,14 +6,27 @@ import * as Types from "measures/CommonQuestions/types";
...
   let ratesExist = false;
   const rateExists = (rate: Types.RateFields) =>
     rate?.rate || (rate?.denominator && rate?.numerator);
+  const PCRADrateExists = (rate: any) => rate?.value;
 
   const performanceMeasureRates = data.PerformanceMeasure?.rates;
   if (performanceMeasureRates) {
     for (const option in performanceMeasureRates) {
+      // Ugly, but PCRAD is the only measure that breaks this check.
+      // If you are thinking about adding another hard-coded value here, consider refactoring this function.
```
Comment (charmcitygavin) [inline services/ui-src/src/utils/form.ts:22]: Nice comment/documentation 👍

**Review by charmcitygavin — APPROVED** [review]
LGTM! Added some questions for my own curiosity

File: services/ui-src/src/components/MultiRate/index.tsx [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+
+  const { field } = useController({
+    name,
+    control,
+    defaultValue: rates,
+  });
+
+  // This is the basic structure of the NDR sets.
+  // There are 4 rates, each is calculated by a combination of the other fields.
+  const ndrForumlas = [
```
[the remark this answers is not in the document]
Comment (jphechter) [reply services/ui-src/src/components/MultiRate/index.tsx:44]: We still calculate NDRs in this component, but the values are generated slightly differently. Each object in `ndrFormulas` is a key to generate a rate. For the first one, the numerator is the 2nd field, the denominator is the first, and the rate is the third.

File: services/ui-src/src/components/MultiRate/index.tsx [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+    if (!allNumbers.test(newValue)) return;
+
+    const digitsAfterDecimal = 4;
+    const prevRate = [...field.value];
+    const editRate = { ...prevRate[index] };
+
+    let validValue;
+
+    // TODO: Need clarification on input validations - Figma says both 10 and 16 digits for inputs
+    // Rates need 4 decimal places, but is that included in 16 digits?
```
[the remark this answers is not in the document]
Comment (jphechter) [reply services/ui-src/src/components/MultiRate/index.tsx:102]: I don't know if I did. Probably worth talking to Danny again on this.

This is why I leave myself notes... so others can remind me that I once had an important thought.

[the remark this answers is not in the document]
Comment (jphechter) [reply services/ui-src/src/components/MultiRate/index.tsx:102]: Just talked to Danny. We're good on this. Removing the comment.

**Comment by abelzer212** [comment]
Screen Shot 2022-03-15 at 1 39 03 PM
Change readmissinos to readmissions

**Comment by abelzer212** [comment]
Screen Shot 2022-03-15 at 1 40 07 PM
Should appear blank if zero is entered, not NAN

**Comment by abelzer212** [comment]
Screen Shot 2022-03-15 at 1 41 29 PM
Rate shouldn't show up as infinity, it should be blank

**Comment by abelzer212** [comment]
No validation appearing at the bottom of the screen if a denominator is entered as zero

**Comment by abelzer212** [comment]
Screen Shot 2022-03-15 at 1 43 30 PM
There's no margin on these three errors and there needs to be

**Comment by abelzer212** [comment]
Screen Shot 2022-03-15 at 1 45 00 PM
Data filled in but OMS section doesn't appear
Screen Shot 2022-03-15 at 1 45 05 PM

**Comment by abelzer212** [comment]
I actually can't get the OMS section to show up at all

**Review by charmcitygavin — CHANGES_REQUESTED** [review]
Other than the minor typo and a [failing test](https://github.com/CMSgov/cms-mdct-qmr/runs/5571980935?check_suite_focus=true) (looks like a string mismatch), looks great! 👍🏻

File: services/ui-src/src/components/MultiRate/index.test.tsx [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+import { renderWithHookForm } from "utils/testUtils/reactHookFormRenderer";
+
+import { MultiRate } from ".";
+
+// Example input fields
+const qualifiers = [
+  "Count of Index Hospital Stays",
+  "Count of Observed 30-Day Readmissions",
+  "Observed Readmission Rate",
+  "Count of Expected 30-Day Readmissinos",
```
[the remark this answers is not in the document]
Comment (jphechter) [reply services/ui-src/src/components/MultiRate/index.test.tsx:13]: Nice catch!

**Review by abelzer212 — DISMISSED** [review]

**Review by charmcitygavin — APPROVED** [review]
👍🏻  🎉

**Review by abelzer212 — DISMISSED** [review]

**Review by abelzer212 — APPROVED** [review]

**Pull request merged**